src/replication.c

```diff
@@ -2607,135 +2730,33 @@ static void fullSyncWithPrimary(connection *conn) {
-        return;
-    }
-    /* Receive end offset response */
-    if (server.repl_rdb_channel_state == REPL_DUAL_CHANNEL_RECEIVE_ENDOFF) {
-        uint64_t rdb_client_id;
-        err = receiveSynchronousResponse(conn);
-        if (err == NULL) goto error;
-        if (err[0] == '\0') {
-            /* Retry again later */
-            serverLog(LL_DEBUG, "Received empty $ENDOFF response");
-            sdsfree(err);
-            return;
-        }
-        long long reploffset;
-        char primary_replid[CONFIG_RUN_ID_SIZE + 1];
-        int dbid;
-        /* Parse end offset response */
-        char *endoff_format = "$ENDOFF:%lld %40s %d %llu";
-        if (sscanf(err, endoff_format, &reploffset, primary_replid, &dbid, &rdb_client_id) != 4) {
-            serverLog(LL_WARNING, "Received unexpected $ENDOFF response: %s", err);
-            goto error;
-        }
-        sdsfree(err);
-        server.rdb_client_id = rdb_client_id;
-        server.primary_initial_offset = reploffset;
-
-        /* Initiate repl_provisional_primary to act as this replica temp primary until RDB is loaded */
-        server.repl_provisional_primary.conn = server.repl_transfer_s;
-        memcpy(server.repl_provisional_primary.replid, primary_replid, CONFIG_RUN_ID_SIZE);
-        server.repl_provisional_primary.reploff = reploffset;
-        server.repl_provisional_primary.read_reploff = reploffset;
-        server.repl_provisional_primary.dbid = dbid;
-
-        /* Now that we have the snapshot end-offset, we can ask for psync from that offset. Prepare the
-         * main connection accordingly.*/
-        server.repl_transfer_s->state = CONN_STATE_CONNECTED;
-        server.repl_state = REPL_STATE_SEND_HANDSHAKE;
-        serverAssert(connSetReadHandler(server.repl_transfer_s, setupMainConnForPsync) != C_ERR);
-        setupMainConnForPsync(server.repl_transfer_s);
-
-        /* As the next block we will receive using this connection is the rdb, we need to prepare
-         * the connection accordingly */
-        serverAssert(connSetReadHandler(server.repl_rdb_transfer_s, readSyncBulkPayload) != C_ERR);
-        server.repl_transfer_size = -1;
-        server.repl_transfer_read = 0;
-        server.repl_transfer_last_fsync_off = 0;
-        server.repl_transfer_lastio = server.unixtime;
-
-        server.repl_rdb_channel_state = REPL_DUAL_CHANNEL_RDB_LOAD;
-        return;
+        /* fall through */
+    case REPL_DUAL_CHANNEL_RECEIVE_REPLCONF_REPLY:
+        ret = dualChannelReplHandleReplconfReply(conn, &err);
+        if (ret == C_OK) server.repl_rdb_channel_state = REPL_DUAL_CHANNEL_RECEIVE_ENDOFF;
+        break;
+    case REPL_DUAL_CHANNEL_RECEIVE_ENDOFF:
+        ret = dualChannelReplHandleEndOffsetResponse(conn, &err);
+        if (ret == C_OK) server.repl_rdb_channel_state = REPL_DUAL_CHANNEL_RDB_LOAD;
+        break;
+    default: break;
     }
+    sdsfree(err);
+    if (ret == C_ERR) goto error;
```

Comment:
I think we should give the error handling logic a chance to examine the err string. whether or not the error handling logic needs to consume the string is a separate concern. 
```suggestion
    if (ret == C_ERR) goto error;
    sdsfree(err);
```

Reply:
Right so we should change the order to
```
    if (ret == C_ERR) goto error;
    sdsfree(err);
```
Also maybe we better log the error string here as well just in case we didn't log it earlier (better log twice then miss it).